scylla/tests/integration/caching_session.rs

```diff
@@ -0,0 +1,140 @@
+            let mut batch = Batch::new(BatchType::Logged);
+            batch.append_statement(insert_1_b_c);
+            batch.append_statement(insert_2_b_c);
+            batch.append_statement(insert_3_b_c);
+            batch.append_statement(insert_4_b_c);
+            // First batch that should generate prepares for each shard.
+            let batch_values = vec![(1, 1), (2, 2), (3, 3), (4, 4)];
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            // Few extra runs. Those batches should not result in any prepares being sent.
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+
+            let prepared_batch_res_rows: Vec<(i32, i32, i32)> = caching_session
+                .execute_unpaged("SELECT * FROM tab", &[])
+                .await
+                .unwrap()
+                .into_rows_result()
+                .unwrap()
+                .rows()
+                .unwrap()
+                .map(|r| r.unwrap())
+                .collect();
+
+
+            // Verify the data from inserts
+            let mut prepared_batch_res_rows = prepared_batch_res_rows;
+            prepared_batch_res_rows.sort();
+            let expected_rows = vec![(1, 1, 1), (2, 2, 2), (3, 3, 3), (4, 4, 4)];
+            assert_eq!(prepared_batch_res_rows, expected_rows);
+
+
+            let feedbacks: usize = feedback_rxs.iter_mut().map(consume_current_feedbacks).sum();
+
+            let expected_feedbacks = (4 + 1) * nr_shards.get() as usize * 3; // 4 inserts + 1 select, 3 nodes
+            assert_eq!(feedbacks, expected_feedbacks);
+
+
+            // Run some alters to invalidate the server side cache, similarly to scylla/src/session_test.rs
+            caching_session.ddl("ALTER TABLE tab RENAME c to tmp").await.unwrap();
+            caching_session.ddl("ALTER TABLE tab RENAME b to c").await.unwrap();
+            caching_session.ddl("ALTER TABLE tab RENAME tmp to b").await.unwrap();
+
+            // execute_unpageds caused by alters likely resulted in some prepares being sent.
+            // Consume those frames.
+            feedback_rxs.iter_mut().map(consume_current_feedbacks).sum::<usize>();
+
+            // Run batches many times per node to ensure we hit every one of them.
+            // Server side cache was invalidated, but the reprepare will be required only if
+            // we hit the node that had its cache invalidated. Only the corresponding shards
+            // will be targeted, so this time expected number of prepares is not multiplied by nr_shards
+            for _ in 0..(64*3) {
+                caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            }
```

Comment:
Instead of making a lot of queries (which can actually fail to reach all shards if you are unlucky) you can use a custom load balancing policy to send a query to each shard exactly once. See tablets.rs integration tests and `SingleTargetLBP` inside it (it is used in `send_statement_everywhere` and `send_unprepared_query_everywhere` helper functions).

Reply:
Applied same logic as in `send_statement_everywhere` but in a synchronous manner. Moved some code from `tablets.rs` to `common/utils.rs` in a separate commit to use it.